processing/src/main/java/io/druid/query/groupby/epinephelinae/AbstractBufferGrouper.java

```diff
@@ -0,0 +1,196 @@
+   * can use this to update its own state, e.g. tracking bucket offsets in a structure outside of the hash table.
+   *
+   * @param bucketOffset offset of the new bucket, within the buffer returned by hashTable.getTableBuffer()
+   */
+  public abstract void newBucketHook(int bucketOffset);
+
+  /**
+   * Called to check if it's possible to skip aggregation for a row.
+   *
+   * @param bucketWasUsed Was the row a new entry in the hash table?
+   * @param bucketOffset Offset of the bucket containing this row's entry in the hash table,
+   *                     within the buffer returned by hashTable.getTableBuffer()
+   * @return true if aggregation can be skipped, false otherwise.
+   */
+  public abstract boolean canSkipAggregate(boolean bucketWasUsed, int bucketOffset);
+
+  /**
+   * Called after a row is aggregated. An implementing BufferGrouper class can use this to update
+   * its own state, e.g. reading the new aggregated values for the row's key and acting on that information.
+   *
+   * @param bucketOffset Offset of the bucket containing the row that was aggregated,
+   *                     within the buffer returned by hashTable.getTableBuffer()
+   */
+  public abstract void afterAggregateHook(int bucketOffset);
+
+  // how many times the hash table's buffer has filled/readjusted (through adjustTableWhenFull())
+  public int getGrowthCount()
+  {
+    return hashTable.getGrowthCount();
+  }
+
+  // Number of elements in the table right now
+  public int getSize()
+  {
+    return hashTable.getSize();
+  }
+
+  // Current number of available/used buckets in the table
+  public int getBuckets()
+  {
+    return hashTable.getBuckets();
+  }
+
+  // Maximum number of elements in the table before it must be resized
+  public int getMaxSize()
+  {
+    return hashTable.getMaxSize();
+  }
+
+  @Override
+  public boolean aggregate(KeyType key, int keyHash)
+  {
+    final ByteBuffer keyBuffer = keySerde.toByteBuffer(key);
+    if (keyBuffer == null) {
+      return false;
+    }
+
+    if (keyBuffer.remaining() != keySize) {
+      throw new IAE(
+          "keySerde.toByteBuffer(key).remaining[%s] != keySerde.keySize[%s], buffer was the wrong size?!",
+          keyBuffer.remaining(),
+          keySize
+      );
+    }
+
+    // find and try to expand if table is full and find again
+    int bucket = hashTable.findBucketWithAutoGrowth(keyBuffer, keyHash);
+    if (bucket < 0) {
+      return false;
+    }
+
+    final int bucketStartOffset = hashTable.getOffsetForBucket(bucket);
+    final boolean bucketWasUsed = hashTable.isBucketUsed(bucket);
+    final ByteBuffer tableBuffer = hashTable.getTableBuffer();
+
+    // Set up key and initialize the aggs if this is a new bucket.
+    if (!bucketWasUsed) {
+      hashTable.initializeNewBucketKey(bucket, keyBuffer, keyHash);
+      for (int i = 0; i < aggregators.length; i++) {
+        aggregators[i].init(tableBuffer, bucketStartOffset + aggregatorOffsets[i]);
+      }
+
```

Comment:
I went with this approach since assumptions about the key+hash layout in the bucket are built into how the hash table operates, but the aggregators are in the "value" portion of a bucket which I view as being opaque to the hash table implementation